src/main/kotlin/org/rust/ide/inspections/RustDeprecatedInspection.kt

```diff
@@ -0,0 +1,106 @@
+package org.rust.ide.inspections
+
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.resolve.ref.RustReference
+
+/**
+ * Detects usage of deprecated API
+ */
+class RustDeprecatedInspection : RustLocalInspectionTool() {
+    override fun getDisplayName(): String = "Deprecated API usage"
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean): PsiElementVisitor =
+        object : RustElementVisitor() {
+            override fun visitMethodCallExpr(el: RustMethodCallExprElement) = inspectReference(el.reference, el.identifier, holder)
+            override fun visitPathExpr(el: RustPathExprElement) = inspectReference(el.path.reference, el, holder)
+            override fun visitPathType(el: RustPathTypeElement) = inspectReference(el.path?.reference, el.path, holder)
+            override fun visitStructExpr(el: RustStructExprElement) = inspectReference(el.path.reference, el.path, holder)
+            override fun visitStructExprField(el: RustStructExprFieldElement) = inspectReference(el.reference, el.identifier, holder)
+            override fun visitTraitRef(el: RustTraitRefElement) = inspectReference(el.path.reference, el, holder)
+            override fun visitUseGlob(el: RustUseGlobElement) = inspectReference(el.reference, el, holder)
+            override fun visitUseItem(el: RustUseItemElement) {
+                if (el.useGlobList == null) {
+                    inspectReference(el.path?.reference, el.path, holder)
+                }
+            }
+        }
+
+    fun inspectReference(ref: RustReference?, el: PsiElement?, holder: ProblemsHolder) {
+        if (ref == null || el == null) return
+        val deprInfo = ref.resolve()?.findDeprecation() ?: return
+        var msg = "${deprInfo.element.getTitle()} is deprecated"
+        if (deprInfo.since != null) {
+            msg += " since ${deprInfo.since}"
+        }
+        if (deprInfo.reason != null) {
+            msg += ": ${deprInfo.reason}"
+        }
+        if (deprInfo.note != null) {
+            msg += ": ${deprInfo.note}"
+        }
+        holder.registerProblem(el, msg, ProblemHighlightType.LIKE_DEPRECATED)
+    }
+
+    /**
+     * Detects if element is deprecated and returns the title of the exact tree node
+     * that causes deprecation.
+     */
+    private fun PsiElement.findDeprecation(): DeprecationInfo? {
+        var attr: RustAttrElement? = null
+        if (this is RustOuterAttributeOwner) {
+            attr = findOuterAttr("deprecated") ?: findOuterAttr("rustc_deprecated")
+        }
+        if (attr == null && this is RustInnerAttributeOwner) {
+            attr = findInnerAttr("deprecated") ?: findInnerAttr("rustc_deprecated")
+        }
+        if (attr == null && parent != null) {
+            return parent.findDeprecation()
```

Comment:
>Could you, please, clarify on this part? 

Sure! If module `m` is deprecated, and you use a function `foo` from the module, then you either:

* use a full path to the item: `m::foo`,

* import an item from a module `use m::{foo, bar}`.

That is, if you have a reference to a function, you must have a reference to containing module as well, so it makes sense to warn about the module itself. 

Also keep in mind that the following code compiles without warnings in rutsc:

```
#[deprecated]
pub mod m {
    pub fn foo() {}  
    pub fn bar() { 
        foo() // no deprecation warning here from rustc
    }
}
```